src/optimizer/statistics/stats_storage.cpp

```diff
@@ -6,39 +6,99 @@
 #include "loggers/optimizer_logger.h"
 
 namespace noisepage::optimizer {
-common::ManagedPointer<TableStats> StatsStorage::GetTableStats(catalog::db_oid_t database_id,
-                                                               catalog::table_oid_t table_id) {
-  StatsStorageKey stats_storage_key = std::make_pair(database_id, table_id);
+
+StatsStorageValue &StatsStorage::GetStatsStorageValue(const catalog::db_oid_t database_id,
+                                                      const catalog::table_oid_t table_id,
+                                                      catalog::CatalogAccessor *accessor) {
+  stats_latch_.LockShared();
+
+  StatsStorageKey stats_storage_key{database_id, table_id};
   auto table_it = table_stats_storage_.find(stats_storage_key);
 
-  if (table_it != table_stats_storage_.end()) {
-    return common::ManagedPointer<TableStats>(table_it->second);
+  if (table_it == table_stats_storage_.end()) {
+    stats_latch_.UnlockShared();
+    InsertTableStats(database_id, table_id, accessor);
+    table_it = table_stats_storage_.find(stats_storage_key);
+    stats_latch_.LockShared();
```

Comment:
I think you need to move this one line up?

Reply:
Done